tests/net/mod.rs

```diff
@@ -0,0 +1,782 @@
+    }
+}
+
+/// Mapping from node IDs to actual node instances.
+pub type NodeMap<D> = collections::BTreeMap<<D as DistAlgorithm>::NodeUid, Node<D>>;
+
+/// A virtual network message tied to a distributed algorithm.
+pub type NetMessage<D> =
+    NetworkMessage<<D as DistAlgorithm>::Message, <D as DistAlgorithm>::NodeUid>;
+
+/// Process a step.
+///
+/// Expands every message in the step by turning all broadcast-messages into peer-to-peer messages,
+/// and appends them to the network queue. Additionally, saves a copy of each output to the output
+/// buffer of the `sender` node.
+///
+/// At the end, the number of additional messages created by non-faulty nodes will be returned.
+///
+/// # Panics
+///
+/// The function will panic if the `sender` ID is not a valid node ID in `nodes`.
+// This function is defined outside `VirtualNet` and takes arguments "piecewise" to work around
+// borrow-checker restrictions.
+fn process_step<'a, D>(
+    nodes: &'a mut collections::BTreeMap<D::NodeUid, Node<D>>,
+    sender: D::NodeUid,
+    step: &Step<D>,
+    dest: &mut collections::VecDeque<NetMessage<D>>,
+) -> usize
+where
+    D: DistAlgorithm + 'a,
+    D::Message: Clone,
+    D::Output: Clone,
+{
+    // For non-faulty nodes, we count the number of messages.
+    let faulty = nodes
+        .get(&sender)
+        .expect("Trying to process a step with non-existing node ID")
+        .is_faulty();
+    let mut message_count: usize = 0;
+
+    // Queue all messages for processing.
+    for tmsg in step.messages.iter() {
+        match &tmsg.target {
+            // Single target message.
+            messaging::Target::Node(to) => {
+                if !faulty {
+                    message_count = message_count.saturating_add(1);
+                }
+
+                dest.push_back(NetworkMessage::new(
+                    sender.clone(),
+                    tmsg.message.clone(),
+                    to.clone(),
+                ));
+            }
+            // Broadcast messages get expanded into multiple direct messages.
+            messaging::Target::All => for to in nodes.keys() {
+                if *to == sender {
+                    continue;
+                }
+
+                if !faulty {
+                    message_count = message_count.saturating_add(1);
+                }
+
+                dest.push_back(NetworkMessage::new(
+                    sender.clone(),
+                    tmsg.message.clone(),
+                    to.clone(),
+                ));
+            },
+        }
+    }
+
+    // Collect all outputs (not required for network operation) as a convenience for the user.
+    nodes
+        .get_mut(&sender)
+        .expect("Trying to process a step with non-existing node ID")
+        .outputs
+        .extend(step.output.iter().cloned());
+
+    message_count
+}
+
+/// New network node construction information.
+///
+/// Helper structure passed to node constructors when constructing virtual networks.
```

Comment:
when ~~constructing~~ building.... just a suggestion to avoid too many constructions!

Reply:
It's even more accurate, since we are using a builder. Awesome!